### PR TITLE
Support Quality Service requests in the java client

### PR DESCRIPTION
**Pull request by AlexPerathoner**
https://code.eu.targomo.com/api-server/quality-service/-/issues/222

Together with https://code.eu.targomo.com/api-server/quality-service/-/merge_requests/245
* adds criterionDefinitions previously declared in quality service
* adds `scores` and `rating/location` requests, responses
* upgrading lombok version from 1.18.2 to 1.18.14 (same as in quality service, needed for `@SuperBuilder` customisation)

### CODE REVIEW

File: src/main/java/com/targomo/client/Constants.java [hunk 1]
```diff
@@ -165,6 +165,38 @@ public class Constants {
     public static final String CRS                                                      = "crs";
     public static final String ROUTE_FROM_CENTROID                                      = "routeFromCentroid";
 
+    public static final String LOCATIONS                                                = "locations";
+    public static final String COMPETITORS                                              = "competitors";
+    public static final String CRITERIA                                                 = "criteria";
```
Comment (HLucasTargomo) [inline src/main/java/com/targomo/client/Constants.java:170]: I would also put these under the '// quality constants' section

File: src/main/java/com/targomo/client/api/quality/Location.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @JsonIgnore
+    public boolean point() {
+        return geometry == null;
+    }
+
+    public String computeCacheId(String apiKey, String ratingId) {
+        String pointGeomSuffix = point() ? "POINT" : "GEOMETRY";
+        return apiKey + CACHE_KEY_SEPARATOR + ratingId + CACHE_KEY_SEPARATOR + String.format("%.6f", lat) +
+                CACHE_KEY_SEPARATOR + String.format("%.6f", lng) + CACHE_KEY_SEPARATOR + pointGeomSuffix;
+    }
```
Comment (HLucasTargomo) [inline src/main/java/com/targomo/client/api/quality/Location.java:51]: this shouldn't be in the client library.
this is only used in the quality service itself
same for the 'point' method

File: src/main/java/com/targomo/client/api/quality/criterion/DistanceInfluence.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.targomo.client.api.quality.criterion;
+
+/**
+ * Defines how the distance/time will influences on the score
+ * @author Hugo Lucas
+ */
+public enum DistanceInfluence {
+    NONE("none"),
+    QUADRATIC("quadratic");
+    //TODO: LINEAR("linear")
```
Comment (HLucasTargomo) [inline src/main/java/com/targomo/client/api/quality/criterion/DistanceInfluence.java:10]: you can remove this now I think (as people can change the distance exponent)

File: src/main/java/com/targomo/client/api/quality/criterion/CriterionType.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    MATH_AGGREGATION(Constants.CRITERION_TYPE_MATH_AGGREGATION, SourceType.MATH_AGGREGATION, DistanceInfluence.NONE, EntityAggregation.SUM),
+    STAYPOINT_COUNT(Constants.CRITERION_TYPE_STAYPOINT_COUNT, SourceType.MOBILITY, DistanceInfluence.NONE, EntityAggregation.SUM),
+    TRANSIT_STOPS_SUM(Constants.CRITERION_TYPE_TRANSIT_STOPS_SUM, SourceType.TRANSIT_STOPS, DistanceInfluence.NONE, EntityAggregation.SUM),
+    TRANSIT_STOPS_DISTANCE(Constants.CRITERION_TYPE_TRANSIT_STOPS_DISTANCE, SourceType.TRANSIT_STOPS, DistanceInfluence.QUADRATIC, EntityAggregation.SUM),
+    CRITERION_REFERENCE(Constants.CRITERION_TYPE_REFERENCE, null, null, null);
+
+    private final String name;
+    private final SourceType sourceType;
+    private final DistanceInfluence distanceInfluence;
+    private final EntityAggregation entityAggregation;
```
Comment (HLucasTargomo) [inline src/main/java/com/targomo/client/api/quality/criterion/CriterionType.java:32]: I think in the client, we don't need to have sourceType, distanceInfluence,and entity Aggregation

File: pom.xml [hunk 5]
```diff
@@ -86,7 +86,7 @@
         <dependency>
             <groupId>org.projectlombok</groupId>
             <artifactId>lombok</artifactId>
-            <version>1.18.2</version>
+            <version>1.18.14</version>
```
Comment (hudda10) [inline pom.xml:89]: @AlexPerathoner  please consider 
https://github.com/projectlombok/lombok/issues/2599
https://github.com/projectlombok/lombok/issues/2607

1.18.14 has some breaking changes.

Comment (HLucasTargomo) [reply pom.xml:89]: After discussing with hudda10, we decided to keep version 1.18.14.
As I'm using this lombok version on quality service for a while now and we never add problems
The issues seemed related to eclipse so not relevant for our use case
If we see any lombok issue in the next days while testing we could try a higher version

**Review by hudda10 — APPROVED** [review]

**Pull request merged**